flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkJoinToMultiJoinRule.java

```diff
@@ -0,0 +1,552 @@
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.rex.RexVisitorImpl;
+import org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.calcite.util.ImmutableIntList;
+import org.apache.calcite.util.Pair;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Flink Planner rule to flatten a tree of {@link LogicalJoin}s into a single {@link MultiJoin} with
+ * N inputs.
+ *
+ * <p>An input is not flattened if the input is a null generating input in an outer join, i.e.,
+ * either input in a full outer join, the right hand side of a left outer join, or the left hand
+ * side of a right outer join.
+ *
+ * <p>Join conditions are also pulled up from the inputs into the topmost {@link MultiJoin}, unless
+ * the input corresponds to a null generating input in an outer join,
+ *
+ * <p>Outer join information is also stored in the {@link MultiJoin}. A boolean flag indicates if
+ * the join is a full outer join, and in the case of left and right outer joins, the join type and
+ * outer join conditions are stored in arrays in the {@link MultiJoin}. This outer join information
+ * is associated with the null generating input in the outer join. So, in the case of a left outer
+ * join between A and B, the information is associated with B, not A.
+ *
+ * <p>Here are examples of the {@link MultiJoin}s constructed after this rule has been applied on
+ * following join trees.
+ *
+ * <ul>
+ *   <li>A JOIN B &rarr; MJ(A, B)
+ *   <li>A JOIN B JOIN C &rarr; MJ(A, B, C)
+ *   <li>A LEFT JOIN B &rarr; MJ(A, B), left outer join on input#1
+ *   <li>A RIGHT JOIN B &rarr; MJ(A, B), right outer join on input#0
+ *   <li>A FULL JOIN B &rarr; MJ[full](A, B)
+ *   <li>A LEFT JOIN (B JOIN C) &rarr; MJ(A, MJ(B, C))), left outer join on input#1 in the outermost
+ *       MultiJoin
+ *   <li>(A JOIN B) LEFT JOIN C &rarr; MJ(A, B, C), left outer join on input#2
+ *   <li>(A LEFT JOIN B) JOIN C &rarr; MJ(MJ(A, B), C), left outer join on input#1 of the inner
+ *       MultiJoin TODO
+ *   <li>A LEFT JOIN (B FULL JOIN C) &rarr; MJ(A, MJ[full](B, C)), left outer join on input#1 in the
+ *       outermost MultiJoin
+ *   <li>(A LEFT JOIN B) FULL JOIN (C RIGHT JOIN D) &rarr; MJ[full](MJ(A, B), MJ(C, D)), left outer
+ *       join on input #1 in the first inner MultiJoin and right outer join on input#0 in the second
+ *       inner MultiJoin
+ * </ul>
+ *
+ * <p>The constructor is parameterized to allow any sub-class of {@link Join}, not just {@link
+ * LogicalJoin}.
+ *
+ * @see FilterMultiJoinMergeRule
+ * @see ProjectMultiJoinMergeRule
+ * @see CoreRules#JOIN_TO_MULTI_JOIN
+ */
+public class FlinkJoinToMultiJoinRule extends RelRule<FlinkJoinToMultiJoinRule.Config>
```

Comment:
add some comments about the relation between FlinkJoinToMultiJoinRule and JoinToMultiJoinRule

Reply:
> add some comments about the relation between FlinkJoinToMultiJoinRule and JoinToMultiJoinRule, and which part is changed

done!